src/ClientData/CallstackDataTest.cpp

```diff
@@ -200,4 +200,124 @@ TEST(CallstackData, FilterCallstackEventsBasedOnMajorityStartExcludesFunctionToS
+
+  std::vector<CallstackEvent> callstack_list;
+  auto insert_callstack = [&](const CallstackEvent& event) { callstack_list.push_back(event); };
+  const uint32_t kResolution = 2000;
+  callstack_data.ForEachCallstackEventOfTidInTimeRangeDiscretized(kTid, 0, 1000, kResolution,
+                                                                  insert_callstack);
+  EXPECT_THAT(callstack_list, testing::Pointwise(CallstackEventEq(), std::vector<CallstackEvent>{
+                                                                         event1, event2, event4}));
+  callstack_data.ForEachCallstackEventOfTidInTimeRangeDiscretized(kAnotherTid, 0, 1000, kResolution,
+                                                                  insert_callstack);
+  EXPECT_THAT(callstack_list,
+              testing::Pointwise(CallstackEventEq(),
+                                 std::vector<CallstackEvent>{event1, event2, event4, event3}));
+
+  callstack_list.clear();
+  callstack_data.ForEachCallstackEventOfTidInTimeRangeDiscretized(
+      kTid, 0, event2.timestamp_ns() + 1, kResolution, insert_callstack);
+  EXPECT_THAT(callstack_list,
+              testing::Pointwise(CallstackEventEq(), std::vector<CallstackEvent>{event1, event2}));
+
+  // With one pixel on the screen we should only see one event.
+  callstack_list.clear();
+  callstack_data.ForEachCallstackEventOfTidInTimeRangeDiscretized(kTid, 0, 1000, /*resolution=*/1,
+                                                                  insert_callstack);
+  EXPECT_THAT(callstack_list,
+              testing::Pointwise(CallstackEventEq(), std::vector<CallstackEvent>{event1}));
+
+  // When max_timestamp is std::numeric_limits<uint64_t>::max(), each callstack should be draw in
+  // the first pixel, and therefore only one will be visible.
+  callstack_list.clear();
+  callstack_data.ForEachCallstackEventOfTidInTimeRangeDiscretized(
+      kTid, 0, std::numeric_limits<uint64_t>::max(), kResolution, insert_callstack);
+  EXPECT_THAT(callstack_list,
+              testing::Pointwise(CallstackEventEq(), std::vector<CallstackEvent>{event1}));
+}
+
+TEST(CallstackData, ForEachCallstackEventInTimeRangeDiscretized) {
+  CallstackData callstack_data;
+
+  const uint64_t cs_id = 12;
+  CallstackInfo cs{{0x11, 0x10}, CallstackType::kComplete};
+  callstack_data.AddUniqueCallstack(cs_id, std::move(cs));
+
+  constexpr uint32_t kTid = 42;
+  constexpr uint32_t kAnotherTid = 43;
+
+  const uint64_t time1 = 142;
+  CallstackEvent event1{time1, cs_id, kTid};
+  callstack_data.AddCallstackEvent(event1);
+
+  const uint64_t time2 = 242;
+  CallstackEvent event2{time2, cs_id, kTid};
+  callstack_data.AddCallstackEvent(event2);
+
+  const uint64_t time3 = 342;
+  CallstackEvent event3{time3, cs_id, kAnotherTid};
+  callstack_data.AddCallstackEvent(event3);
+
+  const uint64_t time4 = 442;
+  CallstackEvent event4{time4, cs_id, kTid};
+  callstack_data.AddCallstackEvent(event4);
+
+  std::vector<CallstackEvent> callstack_list;
```

Comment:
Is this a copy-paste from the code above? Let's make these variables global.

Reply:
The variables still seem to be cope-pasted, but you've given a thumb up. Have you missed a commit?

Reply:
I didn't touch the previous existent tests too much and I've just modified the new cases where the callstack_id doesn't matter.

TBH, I didn't take a real look to the previous existent tests. I was waiting for approval of this PR, and I can try to do something similar for the other cases, although the events differ (no fully repeated code). I can make it in a following PR.

Reply:
OK.